gms/gossiper.cc

```diff
@@ -1701,13 +1729,42 @@ void gossiper::apply_new_states(inet_address addr, endpoint_state& local_state,
     utils::chunked_vector<application_state> changed;
     auto&& remote_map = remote_state.get_application_state_map();
 
+    bool replicate_failed = false;
+    bool listener_failed = false;
+
+    auto isolate_node = [&] {
+        try {
+            logger.warn("Node {} is isolated by gossip because it failed to run gossip listener. A rolling restart is needed to recover!", addr);
+            _isolated_nodes_due_to_listener_failure.insert(addr);
+            container().invoke_on_others([addr] (gossiper& g) {
+                g._isolated_nodes_due_to_listener_failure.insert(addr);
+            }).get();
+            logger.info("Mark node {} as DOWN because it is isloated", addr);
+            convict(addr);
+        } catch (...) {
+            logger.error("Failed to isolate node {}", addr, std::current_exception());
+        }
```

Comment:
Who are we isolating here? The node we got the update for? But it didn't have any problem.  We should isolate ourselves!

A different approach to the problem is to make sure we never ever fail. That requires pre-allocating state for each node when we first add it, and it's not simple.

Reply:
Consider A sends updates to B. B fails to call gossip listeners for A. It could that A has something wrong and sends B something unexpected. It is not necessarily B fails to apply just because memory allocation. 

Currently, B will abort itself. After this patch, B does not abort. Instead, B stops everything with A and marks A as DOWN. Fom B's point of view, node A is down. This has minimum limited impact. The impact is limited with the node A. 

Even if it is B fails to call the listener because of memory allocation. It is the fault of node B, it does not matter. It does not change the fact B failed the listener for A. B either stop everything with A or abort itself,  to prevent B could possibly have a wrong state for A forever.

Reply:
Let's say the cause was bad_alloc. In the current code, the node restarts and will recover. With the new code, A thinks it updated B, but B is in some mixed state with some listeners updated and the others not updated.

Maybe we survive longer, but the total number of states we have to consider is much larger, and so the potential for encountering new bugs is larger.

Reply:
A can not assume anything about B even if A sends something to B with / without this patch.  After this patch, A will think B is down because B does not respond anything with A. B will mark A as DOWN too. The state that A and B mark each other DOWN looks pretty safe, as if they were isolated from network. 

I agree it makes things a bit more complicated, but it can prevent possible ALL nodes down situation which is a total disaster. 

Even if with this patch, we need to notify the user to perform rolling restart to recover the cluster to a health state. The problem of having more total number of states will be gone after rolling restart. In the end, we will have similar results as we abort immediate except there is no abort and no whole cluster down.

Reply:
The problem here is that the number of cases explodes. First we have a failure during gossip update. It's already a rare occurrence and we have no testing for it. And now the rare case survives and can accumulate more rare cases.

I don't agree about the whole cluster going down. If it's a transient bug (memory allocation failure) then there's no reason to expect other nodes will hit it. If it's a software bug, then all nodes will have it and all nodes will start isolating each other.

One exception is during upgrade, if the first upgraded node goes crazy and causes exceptions in all the old nodes, then isolating it is a good choice. But it only works for gossip, if it generates fatal exceptions in other components it doesn't help.

Reply:
> I don't agree about the whole cluster going down. If it's a transient bug (memory allocation failure) then there's no reason to expect other nodes will hit it. If it's a software bug, then all nodes will have it and all nodes will start isolating each other.

Your bad_alloc example will mostly certainly not trigger whole cluster down. However, we saw whole cluster down happened in practice with our user! This is the motivation of this patch as part of the whole fix (fix in generic gossip layer and fix in specific user of gossip) in the first place. It is not that all nodes will start isolating each other. The isolating happens between only existing nodes with the problematic node. 

In the long run, we are moving away more critical usages (e.g., membership ) from gossip, leaving uncritical things (e.g., cache hit rate, loads) only. It is becoming more safe in terms of correctness even when gossip has more cases you worried.

Reply:
I'm still worried that this moves us into untested code. We don't even have a test to exercise this once, and it can happen in many places.

An abort() is bad, yes, but running an untested configuration is also bad.